{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4b896ad5-cf26-43b6-9b1f-2f5cad154fb6",
   "metadata": {},
   "source": [
    "# Running codes in R environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "936adff1-aeed-43d1-ab56-45faec375458",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load R package\n",
    "# 加载R包\n",
    "library(readxl)\n",
    "library(stringr)\n",
    "library(Biostrings)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "library(ggsci)\n",
    "library(patchwork)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f4f4eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read protein sequence in Arabidopsis\n",
    "# 读取拟南芥的蛋白序列\n",
    "protein <- readAAStringSet(\"genome/Athaliana_167_protein_flt.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "532a7d21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read Sumoylation sites\n",
    "# 读取Sumoylation sites信息\n",
    "high <- read_excel(\"Sumoylation from highly sensitive.xlsx\")\n",
    "high <- unique(high)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7a30a88b",
   "metadata": {},
   "outputs": [],
   "source": [
    "high2 <- NULL\n",
    "for(i in 1:nrow(high)){\n",
    "    Gene <- unlist(str_split(high$`Fasta headers`[i],\"\\\\|\"))[1]\n",
    "    Gene <- gsub(\" \",\"\",Gene)\n",
    "    Position <- unlist(str_split(high$`Positions within proteins`[i],\";\"))[1]\n",
    "    re <- data.frame(Gene,Position,stringsAsFactors = F)\n",
    "    high2 <- rbind(high2,re)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "64901e47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read Sumoylation sites\n",
    "# 读取Sumoylation sites信息\n",
    "low <- read_excel(\"Sumoylation from low-throughput studies.xlsx\")\n",
    "proteomic <- read_excel(\"Sumoylation from proteomic studies.xlsx\")\n",
    "low <- rbind(low,proteomic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7b53f46d",
   "metadata": {},
   "outputs": [],
   "source": [
    "low <- low %>%\n",
    "    dplyr::select(Protein,Position=`Positions within proteins`) %>%\n",
    "    unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a2c4ad6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Name <- data.frame(Gene=names(protein))\n",
    "Protein <- lapply(Name$Gene,function(g){\n",
    "    g <- unlist(str_split(g,\"\\\\.\"))[1]\n",
    "})\n",
    "Protein <- unlist(Protein)\n",
    "Name$Protein <- Protein\n",
    "Name <- unique(Name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "df097962",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(Protein)`\n",
      "Warning message in inner_join(., Name):\n",
      "\"\u001b[1m\u001b[22mDetected an unexpected many-to-many relationship between `x` and `y`.\n",
      "\u001b[36mℹ\u001b[39m Row 7 of `x` matches multiple rows in `y`.\n",
      "\u001b[36mℹ\u001b[39m Row 17200 of `y` matches multiple rows in `x`.\n",
      "\u001b[36mℹ\u001b[39m If a many-to-many relationship is expected, set `relationship = \"many-to-many\"` to silence this warning.\"\n"
     ]
    }
   ],
   "source": [
    "low <- low %>%\n",
    "    inner_join(Name) %>%\n",
    "    dplyr::select(Gene,Position) %>%\n",
    "    unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6edd842e",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- rbind(high2,low)\n",
    "data <- unique(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5ced0c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check Sumoylation sites\n",
    "# 核对一下Sumoylation sites信息\n",
    "positive <- NULL\n",
    "for(i in 1:nrow(data)){\n",
    "    g <- data$Gene[i]\n",
    "    pro <- as.character(protein[g])\n",
    "    pos <- data$Position[i]\n",
    "    AA <- substr(pro,pos,pos)\n",
    "    if(AA==\"K\"){\n",
    "        re <- data.frame(Gene=g,Position=pos)\n",
    "        positive <- rbind(positive,re)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fd9a8443",
   "metadata": {},
   "outputs": [],
   "source": [
    "positive <- positive %>%\n",
    "    mutate(Name = paste0(Gene,\":\",Position)) %>%\n",
    "    mutate(Type = 'positive') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7176e3f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save(positive,file=\"positive.Rdata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "90ed6737",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2390"
      ],
      "text/latex": [
       "2390"
      ],
      "text/markdown": [
       "2390"
      ],
      "text/plain": [
       "[1] 2390"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(positive)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "155d8f37",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1409902c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract non-sumoylation sites\n",
    "# 提取没有发生sumoylation的赖氨酸位点作为负样本集\n",
    "negative <- lapply(unique(positive$Gene),function(g){\n",
    "    pro <- as.character(protein[g])\n",
    "    aa <- unlist(str_split(pro,\"\"))\n",
    "    pos <- which(aa==\"K\")\n",
    "    Name <- paste0(g,\":\",pos)\n",
    "    re <- data.frame(Gene=g, Position=pos,Name)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6b975444",
   "metadata": {},
   "outputs": [],
   "source": [
    "negative <- do.call('rbind',negative)\n",
    "negative <- unique(negative)\n",
    "negative <- negative %>%\n",
    "    dplyr::filter(!Name %in% positive$Name) %>%\n",
    "    mutate(Type = 'negative')     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "df214d7b-e675-4804-a2cd-70f8e0d63508",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "61342"
      ],
      "text/latex": [
       "61342"
      ],
      "text/markdown": [
       "61342"
      ],
      "text/plain": [
       "[1] 61342"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(negative)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "eb3f713d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Data <- rbind(positive,negative)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "aa099c89",
   "metadata": {},
   "outputs": [],
   "source": [
    "Original <- Data %>%\n",
    "                group_by(Type) %>%\n",
    "                summarise(Number=n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8ac5b564",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Type</th><th scope=col>Number</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>negative</td><td>61342</td></tr>\n",
       "\t<tr><td>positive</td><td> 2390</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Type & Number\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t negative & 61342\\\\\n",
       "\t positive &  2390\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| Type &lt;chr&gt; | Number &lt;int&gt; |\n",
       "|---|---|\n",
       "| negative | 61342 |\n",
       "| positive |  2390 |\n",
       "\n"
      ],
      "text/plain": [
       "  Type     Number\n",
       "1 negative 61342 \n",
       "2 positive  2390 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Original"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf40f8b3-d961-4e04-baf5-538a148c2fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# taking n residues upstream and downstream of each lysine\n",
    "# 提取上下游的序列\n",
    "# n is 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, respectively.\n",
    "# n表示上下游序列的长度\n",
    "for(base_len in seq(16,34,2)){\n",
    "    AA <- c()\n",
    "    for(i in 1:nrow(Data)){\n",
    "        pro <- protein[Data$Gene[i]]\n",
    "        pos <- as.numeric(Data$Position[i])\n",
    "        len <- str_length(as.character(pro))\n",
    "        start <- pos-base_len\n",
    "        end <- pos+base_len\n",
    "        if(start<1){\n",
    "            start_str <- paste(rep(\"X\",abs(start)+1),collapse ='')\n",
    "        }else{\n",
    "            start_str <- \"\"\n",
    "        }\n",
    "        if(end>len){\n",
    "            end_str <- paste(rep(\"X\",end-len),collapse ='')\n",
    "        }else{\n",
    "            end_str <- \"\"\n",
    "        }\n",
    "        aa <- paste0(start_str,substr(pro,start,end),end_str)\n",
    "        aa <- paste0(substr(aa,1,base_len),substr(aa,base_len+2,base_len*2+1))\n",
    "        AA <- c(AA,aa)\n",
    "    }\n",
    "    Data$Sequence <- AA\n",
    "    neg <- subset(Data,Type==\"negative\")\n",
    "    neg2 <- AAStringSet(neg$Sequence)\n",
    "    names(neg2) <- neg$Name\n",
    "    writeXStringSet(neg2,paste0(\"CNN/negative_f\",base_len,\"_for_cd-hit.fa\"))\n",
    "    pos <- subset(Data,Type==\"positive\")\n",
    "    Positive <- pos %>%\n",
    "                dplyr::select(Gene=Name,Sequence) %>%\n",
    "                mutate(Label=\"positive\")\n",
    "    sim <- c(0.4,0.5,0.6,0.7) # sequence similarity\n",
    "    N <- c(2,3,4,5)\n",
    "    for(inx in 1:4){\n",
    "        s <- sim[inx]\n",
    "        # reduce redundant sequence for negative datasets using CD-HIT\n",
    "        # 使用CD-HIT软件进行冗余序列去除\n",
    "        system(paste0(\"cd-hit -i CNN/negative_f\",base_len,\"_for_cd-hit.fa -o CNN/negative_f\",base_len,\n",
    "                      \"_for_cd-hit_c\",s,\"_uniq.fa -c \",s,\" -T 30 -n \",N[inx]))\n",
    "        Negative <- readBStringSet(paste0(\"CNN/\",\"negative_f\",base_len,\"_for_cd-hit_c\",s,\"_uniq.fa\")) \n",
    "        Negative <- data.frame(Gene=names(Negative),Sequence=Negative,Label=\"negative\")\n",
    "        Data2 <- rbind(Positive,Negative)\n",
    "        # train/test split\n",
    "        # 将数据集划分为训练集和测试集\n",
    "        set.seed(123)\n",
    "        inx <- sample(1:nrow(Data2),0.75*nrow(Data2))\n",
    "        train <- Data2[inx,]\n",
    "        test <- Data2[-inx,]\n",
    "        write.csv(train,paste0(\"CNN/f_\",base_len,\"_c\",s,\"_train_positive_negative.csv\"),quote=F,row.names=F)\n",
    "        write.csv(test,paste0(\"CNN/f_\",base_len,\"_c\",s,\"_test_positive_negative.csv\"),quote=F,row.names=F)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6ee6afe-5672-4d06-ae2d-31cf4e5730ca",
   "metadata": {},
   "source": [
    "# Running codes in Python environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fdb510d-fb5e-40c3-a807-dc23c43ca0a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# balance.py\n",
    "# The python script is used to balance positive and negative samples\n",
    "# 使用python脚本进行降采样"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b48dd8d2-e526-4461-b406-b31793fc2254",
   "metadata": {},
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Wed Feb 26 15:58:55 2025\n",
    "@author: jasonxu\n",
    "\"\"\"\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "os.chdir(\"C:\\\\Users\\\\jasonxu\\\\Desktop\\\\PlantDeepSUMO\\\\CNN\")\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from imblearn.under_sampling import NearMiss\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "for f in [16,18,20,22,24,26,28,30,32,34]:\n",
    "    print(f)\n",
    "    for s in [0.4,0.5,0.6,0.7]:\n",
    "        print(s)\n",
    "        for data in ['train','test']:\n",
    "            f1 = \"f_\"+str(f)+\"_c\"+str(s)+\"_\"+data+\"_positive_negative.csv\"\n",
    "            df = pd.read_csv(f1)\n",
    "            encoder = OneHotEncoder()\n",
    "            # 将氨基酸序列转换为数值特征\n",
    "            sequences = df['Sequence'].apply(lambda x: list(x)).tolist()\n",
    "            encoded_sequences = encoder.fit_transform(sequences).toarray()\n",
    "            df_encoded = pd.DataFrame(encoded_sequences)\n",
    "            df_encoded['Label'] = df['Label']\n",
    "            df_encoded.index = df['Gene']\n",
    "            df_encoded.head()\n",
    "            # 将数据分为特征和标签\n",
    "            X = df_encoded.drop('Label', axis=1)\n",
    "            y = df_encoded['Label']\n",
    "            # 创建NearMiss对象\n",
    "            nm = NearMiss()\n",
    "            # 进行降采样\n",
    "            X_resampled, y_resampled = nm.fit_resample(X, y)\n",
    "            # 创建新的平衡数据框\n",
    "            df_resampled = pd.DataFrame(X_resampled, columns=X.columns)\n",
    "            df_resampled['Label'] = y_resampled\n",
    "            # 查看平衡后的数据\n",
    "            print(df_resampled['Label'].value_counts())\n",
    "            final = df.iloc[nm.sample_indices_]\n",
    "            f2 = \"f_\"+str(f)+\"_c\"+str(s)+\"_\"+data+\"_positive_negative_balance.csv\"\n",
    "            final.to_csv(f2,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49c09a50-5ea1-432f-95c8-141474939555",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deb40a89",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
